crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl

```diff
@@ -0,0 +1,168 @@
+
+#ifdef MULTISAMPLED
+    let this_motion_vector = textureLoad(motion_vectors, frag_coords, i32(sample_index)).rg;
+#else
+    let this_motion_vector = textureSample(motion_vectors, texture_sampler, in.uv).rg;
+#endif
+
+#ifdef NO_DEPTH_TEXTURE_SUPPORT
+    let this_depth = 0.0;
+    let depth_supported = false;
+#else
+    let depth_supported = true;
+#ifdef MULTISAMPLED
+    let this_depth = textureLoad(depth, frag_coords, i32(sample_index));
+#else
+    let this_depth = textureSample(depth, texture_sampler, in.uv);
+#endif
+#endif
+    
+    // The exposure vector is the distance that this fragment moved while the camera shutter was
+    // open. This is the motion vector (total distance traveled) multiplied by the shutter angle (a
+    // fraction). In film, the shutter angle is commonly 0.5 or "180 degrees" (out of 360 total).
+    // This means that for a frame time of 20ms, the shutter is only open for 10ms.
+    //
+    // Using a shutter angle larger than 1.0 is non-physical, objects would need to move further
+    // than they physically travelled during a frame, which is not possible. Note: we allow values
+    // larger than 1.0 because it may be desired for artistic reasons.
+    let exposure_vector = shutter_angle * this_motion_vector;
+
+    var accumulator: vec4<f32>;
+    var weight_total = 0.0;
+    let n_samples = i32(settings.samples);
+    let noise = hash_noise(frag_coords, globals.frame_count); // 0 to 1
+       
+    for (var i = -n_samples; i < n_samples; i++) {
+        // The current sample step vector, from in.uv
+        let step_vector = 0.5 * exposure_vector * (f32(i) + noise) / f32(n_samples);
+        var sample_uv = in.uv + step_vector;
+        let sample_coords = vec2<i32>(sample_uv * texture_size);
+
+    #ifdef MULTISAMPLED
+        let sample_color = textureLoad(screen_texture, sample_coords, i32(sample_index));
+    #else
+        let sample_color = textureSample(screen_texture, texture_sampler, sample_uv);
+    #endif
+    #ifdef MULTISAMPLED
+        let sample_motion = textureLoad(motion_vectors, sample_coords, i32(sample_index)).rg;
+    #else
+        let sample_motion = textureSample(motion_vectors, texture_sampler, sample_uv).rg;
+    #endif
+    #ifdef NO_DEPTH_TEXTURE_SUPPORT
+        let sample_depth = 0.0;
+    #else
+    #ifdef MULTISAMPLED
+        let sample_depth = textureLoad(depth, sample_coords, i32(sample_index));
+    #else
+        let sample_depth = textureSample(depth, texture_sampler, sample_uv);
+    #endif
+    #endif
+
+        var weight = 1.0;
+        let is_sample_in_fg = !(depth_supported && sample_depth < this_depth && sample_depth > 0.0);
+        if is_sample_in_fg {
+            // The following weight calculation is used to eliminate ghosting artifacts that are
+            // common in motion-vector-based motion blur implementations. While some resources
+            // recommend using depth, I've found that sampling the velocity results in significantly
```

Comment:
Instead of depth, you can try plane distance, which has less artifacts on parallel surfaces. May or may no work, I'm not going to block this PR on it by any means, but if you're interested: https://developer.download.nvidia.com/video/gputechconf/gtc/2020/presentations/s22699-fast-denoising-with-self-stabilizing-recurrent-blurs.pdf#page=44

Also, did you make sure to linearize the depth before doing depth comparisons?

Reply:
I don't understand the connection here, can you add some detail? The depth test is only used to test for occlusion (in front or behind?), I don't see how linearization or the plane distance matters here, I just need a boolean pass/fail.

Reply:
You mentioned there was difficulties accidentally sampling a static foreground object when trying to blur the background. You can compare the depth of each sample to the depth of the pixel you're trying to blur, to have a heuristic to detect different surfaces. It's commonly used to prevent blurring and temporal ghosting across different surfaces in denoisers.

Reply:
> You can compare the depth of each sample to the depth of the pixel you're trying to blur, to have a heuristic to detect different surfaces

Yeah, this is what I spent a ton of time on, and found that depth heuristics are always fragile, because that isn't a good way to seperate objects for blur, but motion similarity is.

Reply:
The problem I found with depth heuristics is that for MB, the most annoying artifact you want to reduce is background bleeding. What you are trying to do with the depth heuristic is figure out if the thing you are sampling could've affected the current pixel, using a depth test is an indirect way to answer the question by answering a _different_ question, which is "is the thing I'm sampling the same object as the thing I'm writing to". Instead, you can answer the _actual_ question by looking at the motion vectors, because it tells you directly whether or not that sample could've been located at the current pixel, based on the speed and direction it is moving.